newlib/libc/tinystdio/fclose.c

```diff
@@ -34,12 +34,27 @@
 int
 fclose(FILE *f)
 {
-        struct __file_close *cf = (struct __file_close *) f;
-        if ((f->flags & __SCLOSE) && cf->close) {
+	int ret = 0;
+#ifdef _WANT_FLOCKFILE
```

Comment:
This doesn't seem right to me -- we want a single function to be responsible for freeing the FILE and cleaning up any locks as those must be cleaned up before the FILE is freed. I think that means pushing the lock cleanup operation down the stack. That seems consistent with all of the other lock usage though; we're holding the lock across calls to the put and get operations.

Reply:
@keith-packard , my opinion on this is that now it has wrong approach, let me explain:

1. memory allocates in [fdopen](https://github.com/picolibc/picolibc/blob/1584f3abaed2f1071060dfd6e9294a46d1db2bad/newlib/libc/tinystdio/fdopen.c#L51)
2. but now it deallocates in one level down (fclose -> [__bufio_close](https://github.com/picolibc/picolibc/blob/1584f3abaed2f1071060dfd6e9294a46d1db2bad/newlib/libc/tinystdio/bufio.c#L311))

I believe free should be done on the same level - in fclose() function